wikiclass/feature_lists/wikidatawiki.py

```diff
@@ -21,6 +32,164 @@ class items:
     """
     HUMAN = 'Q5'
 
+def _process_item_doc(text):
+    if text is not None:
+        return json.loads(text)
+    else:
+        return None
+
+def _process_item(item_doc):
+    item = pywikibase.ItemPage()
+    item.get(content=item_doc or {'aliases': {}})
+    return item
+
+def _process_all_sources(item):
+
+	list_sources_in_JSON = []
+	
+	try:
+		for property in item.claims:
+			list_of_properties = property
+			for claim in item.claims[list_of_properties]:
+				list_of_claims = claim
+				for i, source in enumerate(list_of_claims.sources):
+					for index_list_1, index_list_2 in source.items(): 
+						sources_in_JSON = index_list_2[0].toJSON()
+						list_sources_in_JSON.append(sources_in_JSON['datavalue']['value'])
+		
+		return list_sources_in_JSON
+	except:
+		return 0 #if the revision does not contain any sources, return 0
+
+def _process_wikimedia_sources(item):
+
+	list_sources_in_JSON = []
+	list_wikimedia_sources = []
+	
+	
+	for property in item.claims:
+		list_of_properties = property
+		for claim in item.claims[list_of_properties]:
+			list_of_claims = claim
+			for i, source in enumerate(list_of_claims.sources):
+				for index_list_1, index_list_2 in source.items(): 
+					sources_in_JSON = index_list_2[0].toJSON()
+					list_sources_in_JSON.append(sources_in_JSON['datavalue']['value'])
+
+	wdir_path = os.path.dirname(os.path.realpath(__file__)) #get the current working directory path
+	csv_path = os.path.join(wdir_path, 'excluded_qids.csv')
+
+	for list_sources_in_JSON_content in list_sources_in_JSON: 
+		with open(csv_path) as csvfile:
+			readCSV = csv.reader(csvfile, delimiter=',')
+			for line in readCSV:				
+				try:
+					if(list_sources_in_JSON_content['numeric-id'] == int(line[0])): #if a source comes from Wikimedia projects and DBpedia (i.e. DBpedia collects data from Wikipedia), append that source to the list
+						list_wikimedia_sources.append(list_sources_in_JSON_content)
+						break
+				except:
+					continue
+	return list_wikimedia_sources
+
+def _process_external_sources(item):
+    
+    count_all_sources = len(_process_all_sources(item))
+    count_wiki_sources = len(_process_wikimedia_sources(item))
+    return count_all_sources - count_wiki_sources
+
+
+def _process_external_sources_ratio(item):
+	
+	try: 
+		count_external_sources = _process_external_sources(item)
+		count_claims_with_sources = len(_process_sources(item))
```

Comment:
This processing should depend on [source, external_sources] so that work isn't duplicated.